taskcat/testing/manager.py

```diff
@@ -0,0 +1,252 @@
+        :param regions: comma separated list of regions to test in
+        :param skip_upload: Use template in an existing cloudformation bucket.
+        :param lint_disable: disable cfn-lint checks
+        """
+
+        _trim_regions(regions, self.config)
+        _trim_tests(test_names, self.config)
+
+        boto3_cache = Boto3Cache()
+
+        templates = self.config.get_templates()
+
+        if skip_upload and not self.config.config.project.s3_bucket:
+            raise TaskCatException(
+                "cannot skip_buckets without specifying s3_bucket in config"
+            )
+
+        buckets = self.config.get_buckets(boto3_cache)
+
+        if not skip_upload:
+            # 1. lint
+            if not lint_disable:
+                lint = TaskCatLint(self.config, templates)
+                errors = lint.lints[1]
+                lint.output_results()
+                if errors or not lint.passed:
+                    raise TaskCatException("Lint failed with errors")
+            # 2. build lambdas
+            if self.config.config.project.package_lambda:
+                LambdaBuild(self.config, self.config.project_root)
+            # 3. s3 sync
+            stage_in_s3(
+                buckets, self.config.config.project.name, self.config.project_root
+            )
+
+        regions = self.config.get_regions(boto3_cache)
+        parameters = self.config.get_rendered_parameters(buckets, regions, templates)
+        tests = self.config.get_tests(templates, regions, buckets, parameters)
+
+        self.test_definition = Stacker(
+            self.config.config.project.name,
+            tests,
+            shorten_stack_name=self.config.config.project.shorten_stack_name,
+        )
+        self.test_definition.create_stacks()
+        terminal_printer = TerminalPrinter(minimalist=self.minimal_output)
+
+        terminal_printer.report_test_progress(stacker=self.test_definition)
+
+    def end(
+        self,
+        no_delete: bool = False,
+        keep_failed: bool = False,
+        dont_wait_for_delete: bool = True,
+    ):
+        """
+        docstring
+        """
+
+        terminal_printer = TerminalPrinter(minimalist=self.minimal_output)
+
+        status = self.test_definition.status()
+
+        # Delete Stacks
+        if no_delete:
+            LOG.info("Skipping delete due to cli argument")
+        elif keep_failed:
+            if len(status["COMPLETE"]) > 0:
+                LOG.info("deleting successful stacks")
+                self.test_definition.delete_stacks({"status": "CREATE_COMPLETE"})
+        else:
+            self.test_definition.delete_stacks()
+
+        if not dont_wait_for_delete:
+            terminal_printer.report_test_progress(stacker=self.test_definition)
+
+        # TODO: summarise stack statusses (did they complete/delete ok) and print any
+        #  error events
+
+        # Delete Templates and Buckets
+        buckets = self.config.get_buckets()
```

Comment:
I also repeat this line in the start method. I'm not sure what the performance penalty is from getting buckets twice and not passing in a botocache object. 

Should I move buckets to an instance property, botocache, or leave it as is?

Reply:
For the moment I'd like to leave the `Config` class as-is. There are other performance related items we want to address, and I'm not sure what the blast radius would be if we start ad-hoc modifying it. 

2021 will see significant work to improve from current state - much like 2019 did. Go glance at `0.8.50` to see where we came from. 😀

Reply:
I wasn't thinking about the Config class. I need the buckets in both the `start` and `end` methods of `TaskManager`. 

So I'm wondering if I should:

```python
# inside of the start method
self.cache = Boto3Cache()
```
then grab the buckets in each method
```python
# from the end method
buckets = self.config.get_buckets(self.cache)
```

or maybe just save the buckets?

```python
# inside the start method
self.buckets = self.config.get_buckets()
```
Then I can just use `self.buckets` when I need them in the `end` method.

Reply:
I'm following. Caching the buckets in this scenario is acceptable. 

The `config.get_buckets()` method _does_ introspect the Boto3Cache and create new S3-Bucket-Objects on the fly. However that's based on the current-state of the `Config` instantiation. We don't provide an easy way to alter the config in runtime (nor do I have any interest in doing so). The net result is just-in-time objects that are reading from a config that's already in memory. meh.